runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java

```diff
@@ -86,13 +87,17 @@ public UnboundedReadFromBoundedSource(BoundedSource<T> source) {
 
   @Override
   public PCollection<T> expand(PBegin input) {
-    return input.getPipeline().apply(
+    PCollection<T> collection = input.getPipeline().apply(
         Read.from(new BoundedToUnboundedSourceAdapter<>(source)));
+
+    defaultOutputCoder = collection.getCoder();
+
+    return collection;
   }
 
   @Override
   protected Coder<T> getDefaultOutputCoder() {
-    return source.getDefaultOutputCoder();
+    return defaultOutputCoder;
```

Comment:
What if you just don't override this method? Same in other places in this PR.

It seems like getDefaultOutputCoder(), ultimately, is used to provide a Coder for a PCollection produced by a PTransform in case the PTransform didn't already .setCoder() on the PCollection explicitly - see https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/values/PCollection.java#L137 - which means that, for any particular transform, you should be able to remove an override of getDefaultOutputCoder() and instead either call .setCoder() on the PCollection it produces, or even just rely on the other coder inference mechanisms (see code below highlight in the link above).